Skip-List implementation

lembergerth opened the pull request:
Add a [skip list](https://en.wikipedia.org/wiki/Skip_list) implementation with partial counts.

Provides a sorted data structure with a mean performance of
O(log n) for insertion
O(log n) for removal
O(log n) for contains
O(log n) for random access by rank/index
O(1) for first/last item in order

lembergerth committed:
Skip list implementation

kfriedberger commented:
Copyright notice for CPAchecker? Please change!

lembergerth committed:
Remove default copyright in IntelliJ, which was copyright of CPAchecker

lembergerth committed:
Remove wrong copyright headers

kfriedberger commented:
Is your implementation thread-safe?
How does it perform compared to a (Set based on a) [ConcurrentSkipListMap](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListMap.html)?

lembergerth committed:
Refactor skip list: Remove unused code, throw expected expections

lembergerth committed:
Add equals test to SkipListTest that fits SortedSet interface, use generator of correct type

lembergerth committed:
Add sosy-commons copyright to skip list files

lembergerth committed:
Correct copyright license for IntelliJ IDEA

lembergerth commented:
> Copyright notice for CPAchecker? Please change!

This was the default for the existing IntelliJ configuration - I changed it.

> Is your implementation thread-safe?

No.

> How does it perform compared to a (Set based on a) ConcurrentSkipListMap?

My experiments haven't finished yet, but ConcurrentSkipListMap does not provide
random access - otherwise, the collections have the same theoretical background.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.SortedSet;
+
+public interface OrderStatisticSet<T> extends SortedSet<T> {
```
The interface and the methods should have JavaDoc that explains the guarantees.

Why is the name `OrderStatisticSet`? Why not `SortedUniqueList` or something else?

lembergerth replied:
Should I really add performance guarantees to the interface?
Despite that, JavaDoc is obviously missing, of course. I added it.

* Why not `SortedUniqueList`?

Lists provide more methods for indexed access (e.g., to add a value at a specific position),
so I thought that any interface name with `List ` in it may be misleading.
In addition, it would be easily possible to add a *MultiSet and *Map interface and implementation with similar capabilities, so the `Unique` part would have to be removed
and the new interfaces would have to get different names, reducing the 'grouping' of such interfaces.

* Why `OrderStatistic`?

It's not used in any interface description yet, we could easily add `OrderStatisticMultiset` and `OrderStatisticMap` interfaces, and the [meaning](https://en.wikipedia.org/wiki/Order_statistic) is correct: The interface allows 1) to get the k-th smallest value of a set and 2) to determine the k for a given value.
In addition, a data structure [order statistic tree](https://en.wikipedia.org/wiki/Order_statistic_tree) already exists in literature that provides the same capabilities.

That said, I'm open for names that are more self-explanatory, I just couldn't find one yet.

PhilippWendler replied:
Ok, if the name is already well-known, let's use it.

For guarantees in JavaDoc, I meant all kinds of guarantees, but performance would also be interesting if possible and meaningful for the interface (we could look at `Set`, `List`, etc., if they provide performance guarantees). Otherwise only `SkipList` should document performance guarantees.

lembergerth replied:
The collection interfaces don't provide performance guarantees. But could  you have a look at the new JavaDoc to check whether you miss any information, please?

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,31 @@
+
+  T get(int pIndexOf);
+
+  T remove(int pIndexOf);
```
Having `remove(int)` is problematic for `OrderStatisticSet<Integer>`, because calling `remove` with an `Integer` could call both methods. Better rename these two methods.

lembergerth replied:
Good point, I renamed them to {get,remove}ByRank , so that they are still close to the get/remove methods and easily findable.

lembergerth committed:
SkipList provides constructors as expected by SortedSet interface

lembergerth committed:
Add JavaDoc to OrderStatisticSet interface and rename methods to remove ambiguity

kfriedberger commented:
Feature request: Is your implementation `Serializable`?
Most of our data-structures can be serialized.

lembergerth committed:
Make SkipList serializable

lembergerth commented:
Now it is.

lembergerth committed:
Enable serialization of SkipList on large sets.

The default serialization leads to an StackOverflowException
due to the recursive linking of nodes in the set.
Instead, serialize the nodes of the set iteratively.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,113 @@
+ * <p>If the used compare-method is not consistent with <code>equals</code>, unequal (i.e.,
+ * different) elements that are equal according to the order can exist in a set. In this case,
+ * elements that are added first will have a higher rank than elements of the same order that are
+ * added later.
```
Are you sure you want to guarantee a specific behavior if `compareTo` and `equal` are inconsistent? Why not simply require that they are consistent?

Also, this is a contradiction to the paragraph above, where it simply says that implementations "should" work (but are not required to) if the mthods are inconsistent.

lembergerth replied:
Requiring that the used `compare`-method (not just `compareTo`!) is consistent with `equals`  is a strong constraint on the possible uses of the interface.

Whenever you want to use some comparison criterion that is not consistent with `equals`,
you will have to work around this by maintaining collections of objects for each 'comparison value', manually.

Through this definition in the interface, we would allow implementing sets to do this for the user.

> Also, this is a contradiction to the paragraph above, where it simply says that implementations "should" work (but are not required to) if the mthods are inconsistent.

Thanks!

PhilippWendler replied:
`SortedSet` also requires consistency of `equals` and `compareTo`, at least for fulfilling the `Set` contract. We should use the same guarantees and requirements.

`SkipList` doesn't call `equals` at all, right? Then it will work anyway just like users expect it.

lembergerth replied:
Yes, the `SortedSet` interface states:
> […] if the sorted set is to correctly implement the Set interface

So that's equal to our condition, isn't it?

>SkipList doesn't call equals at all, right? Then it will work anyway just like users expect it.

Yes it does. All operations that are specific to a given object (e.g., `remove`, `contains`) use the equals-method to ensure that the objects are not only equal by the given Comparator, but by their own equals-method.

PhilippWendler replied:
> All operations that are specific to a given object (e.g., remove, contains) use the equals-method to ensure that the objects are not only equal by the given Comparator, but by their own equals-method.

This should not be done. Just use only the given order, and document that if `equals` and the given order is inconsistent, then the `Set` contract is not fulfilled. This is the same as all other `SortedSet` implementations I know do.

lembergerth replied:
Oh ok, I can do that.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,113 @@
+   * <p>Example:
+   *
+   * <pre>
+   *     * Element type T: Pair&lt;Integer, Integer&gt;
```
`Pair` is not a good example because it is not defined (who knows how it's `equals` method is implemented?).

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
Is this formatted with `google-java-format`? I prefer annotations for return types on the same line as the return type.

lembergerth replied:
Yes, unfortunately.

> I  prefer annotations for return types on the same line as the return type.

Me too. The way it is formatted, it appears to be an annotation for the whole method.
Is there an easy way to change this?

PhilippWendler replied:
No, we can't change this.

But I noticed that you have many methods that are package-private but could be `private`. If you add the `private` modifier before the `@Nullable`, google-java-format will format as desired.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
```diff
@@ -0,0 +1,756 @@
+ * @param <T> type of the elements of this list
+ * @see java.util.concurrent.ConcurrentSkipListMap
+ */
+public class SkipList<T> implements OrderStatisticSet<T>, Serializable {
```
Should be `final`.

lembergerth replied:
Ok.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
The unchecked cast should be replaced with `Comparator.<? super T>naturalOrder()` (same above).

lembergerth replied:
Will this work?
`T` does not have to extend Comparable, but the type parameter of `Comparator.naturalOrder()` does have to.

There's actually a bigger, more general problem: if no comparator is given and `T` does not implement the Comparable-interface.
But with the two constructors creating `Comparator.naturalOrder()`, there is no way of telling because we do not have an object of Type `T` available.

Looking at the `TreeMap` implementation, they also do not handle this.

PhilippWendler replied:
This is another argument for factory methods, for those that use the natural order you can enfore `T extends Comparable<T>` via the type parameter of the method. Then you don't need the unchecked cast.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
I would suggest factory methods instead of public constructors (Guava-style).

`copyOf` makes it much more clear than `new SkipList(foo)` that this is not a list that contains `foo`. Adding `of` with varargs makes it easier to create lists of a few elements.

Having overloads with `Collection` and `SortedSet` that differ in their semantic regarding which comparator is used is problematic. What if you want a copy of a `SortedSet`, but with a natural order? What if you pass a `SortedSet` instance to the method that accept's a `Collection`? People could expect that it still takes the `SortedSet`'s comparator.

So it is better to use factory methods with distinctive names that make clear with order is used, e.g., `copyOfWithNaturalOrder`, `copyOfWithSameOrder` etc.

If this gets too complex because of too many combinations that we want, we could use a builder instead.

lembergerth replied:
The SortedSet-interface states that implementing methods should provide these four constructors.
If you say that we can disregard that, I'd be happy to add factory methods.

PhilippWendler replied:
For the above reasons, I think that the Guava style is superior.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
In cases like this, it also needs to be chcked whether`pIndex` is negative. There exists `Preconditions.checkElementIndex` precisely for this use case.

lembergerth replied:
Thanks for the hint!

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
The `toString` of collections is pretty much standardized, and should be the same here as well.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
Why implement `ListIterator` if the return type is just `Iterator` anyway?

We could change the return type to `ListIterator`, or expose an additional `listIterator` method like `List` has. But probably exposing a `ListIterator` is not a good idea because it would allow `add` and `set` at a given position.

lembergerth replied:
> We could change the return type to ListIterator, or expose an additional listIterator method like List has. But probably exposing a ListIterator is not a good idea because it would allow add and set at a given position.

Yes, I thought of both, mainly because I implemented the `List`-interface, at first,
but did choose not to do either because of the same reasons.

I just kept the implementation in case it is required at some point, but keeping code lying around is a bad practice, of course. I'll remove it.

PhilippWendler commented on src/org/sosy_lab/common/collect/OrderStatisticSet.java:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @param <T> the type of elements maintained by this set
+ */
+public interface OrderStatisticSet<T> extends SortedSet<T> {
```
Does it make sense for all `OrderStatisticSet`s to extend `NavigableSet`? Or does it at least make sense for `SkipList`?

lembergerth replied:
It definitely makes sense for the `SkipList`.
Order statistic tree implementations are also able to get the floor/ceil values easily,
but I'm not sure whether this property can be derived from the interface, in general.

Do  you have an opinion on that?

PhilippWendler replied:
I know less about these data structures than you, so it is up to you to decide whether this makes sense.

lembergerth replied:
OK, so how about that notion:

Every collection that is able to get the rank of an element in an ordered set has to determine the position of that element in the list.
If it can do that, it can also determine where an element in an ordered set *should* be, and thus determine its neighbors.

Thus, every implementation of `OrderStatisticSet` should also be able to implement `NavigableSet` efficiently.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
If google-java-format does not enforce the current formatting, I would prefer `private @Nullable T value`, because semantically the `@Nullable` is a modifier for the type `T`.

Why does one need to be careful while serializing `null` values?

lembergerth replied:
> If google-java-format does not enforce the current formatting

No it doesn't, your choice makes sense - I changed it.

> Why does one need to be careful while serializing null values?

The comment was an artifact of a poor approach of manual serialization where I read the value as an Object and cast it, I removed it.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
```diff
@@ -0,0 +1,756 @@
+
+/**
+ * A skip list implementation. Similar to ordered trees, but in list form. Achieves time complexity
+ * similar to ordered trees using a probabilistic approach.
```
Does it allow `null` values? This should be documented.

Do we want it to allow `null` values?

lembergerth replied:
No it doesn't.

I think the use of `null` as a value has very few legitimation and would be a special case
regarding the order of elements, too.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipListTest.java:
```diff
@@ -0,0 +1,98 @@
+    @Override
+    protected SortedSet<String> create(String[] pIntegers) {
+      SkipList<String> list = new SkipList<>();
+      // noinspection ResultOfMethodCallIgnored
```
What is this? Why is it necessary?

lembergerth replied:
It removes unnecessary inspections in IntelliJ, I removed it (and another one)

PhilippWendler replied:
It is still there.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipListTest.java:
```diff
@@ -0,0 +1,98 @@
+                CollectionSize.ANY,
+                CollectionFeature.SUPPORTS_ADD,
+                CollectionFeature.SUPPORTS_REMOVE,
+                CollectionFeature.SUPPORTS_ITERATOR_REMOVE,
```
The above three can be replaced with `SetFeature.GENERAL_PURPOSE`.

Probably we also want `CollectionFeature.KNOWN_ORDER`, `CollectionFeature.ALLOWS_NULL_VALUES` (depending on what we decide), and `CollectionFeature.SERIALIZABLE_INCLUDING_VIEWS`.

lembergerth committed:
SKipList random can be reinitialized from outside

lembergerth committed:
SkipList JavDoc states that null-values are not allowed

lembergerth committed:
OrderStatisticSet Javadoc specifies that elements do not have to fulfill that compare is consistent with equals

lembergerth committed:
SkipList.iterator() just returns an Iterator, no ListIterator

lembergerth committed:
Annotation @Nullable comes right before variable type

lembergerth committed:
SkipListTest checks toString() and order of elements

lembergerth committed:
SkipList.toString() returns expected list of elements

lembergerth committed:
SkipList performs more index checks and uses Preconditions.checkElementIndex to do so.

lembergerth committed:
Reformat

lembergerth commented on src/org/sosy_lab/common/collect/SkipList.java:
```diff
@@ -185,6 +185,14 @@ public SkipList() {
+  public void reinitialize(Random pRandom) {
+    Preconditions.checkNotNull(pRandom);
+    randomGenerator = pRandom;
+  }
```
I'd like to point to this method so that it's not overseen:
Do you mind such a method?
The point is to enable the use of a controlled Random-object (with a specific seed)
from outside.

PhilippWendler replied:
First, if the `Random` instance is static as it was before, it should probably be a `ThreadLocalRandom`.

Second, which use cases would need changing the random-number generator during the lifetime of a `SkipList` object, and would not be possible if the random-number generator could only be specified at instantation time?

Third, which use cases actually need specifying the random-number generator?
I can only think of cases where users wish to make their code deterministic, e.g., for tests.
However, I see no reason not to make `SkipList` fully deterministic by default.

Thus, I suggest to give each `SkipList` instance its own `Random` instance created with a constant seed.

lembergerth committed:
Refactor SkipList

lembergerth committed:
SkipList implements NavigableSet

lembergerth committed:
SkipList comparator is null if natural ordering of element is used, to naturally comply to #comparator() contract that it returns null for natural ordering

lembergerth committed:
SkipList iterators as inner classes, make all inner classes of SkipList static for serialization

lembergerth committed:
Fix bugs in SkipList regarding wrong list-tail tracking and null-value handling

lembergerth committed:
Merge branch 'master' of github.com:sosy-lab/java-common-lib

lembergerth committed:
SkipList only uses comparator to determine equality, not real #equals method of objects

lembergerth committed:
OrderStatisticSet does not specify behavior if compare method is not consistent with equals

lembergerth committed:
SkipList uses factory methods

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
```diff
@@ -163,37 +166,86 @@ private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFound
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> create() {
```
Why not name this `of` in the Guava style?

lembergerth replied:
I used the [Guava Wiki](https://github.com/google/guava/wiki/CollectionUtilitiesExplained#static-constructors) as reference,
which states that newer Collection classes use `create()`.

I also find it more speaking than an `of` without parameters, and it can be easily found.

PhilippWendler replied:
This wiki page uses `create()` only as an example and does not say that newer collection classes use `create()`. However, mutable collections might be a better place to look at for reference than immutable collections if ours is mutable. This would indicate to have `create()` and no other `of` methods.

lembergerth replied:
Personally, I don't have any use case where you would need a sorted set of only `n` elements,
with `n` being that small that you write it down by hand (except for testing).
Thus, I don't require the `of` methods, but maybe you have a good use case?

PhilippWendler replied:
Then let's remove it.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
```diff
@@ -163,37 +166,86 @@ private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFound
+    return new SkipList<>(pCollection, pComparator);
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> createWithNaturalOrder(
```
How about `copyOfWithNaturalOrder`?

lembergerth replied:
I had that first, but 'copy' implies that it's copied - and is a collection really copied if the position of its elements could be completely changed?

PhilippWendler replied:
That's maybe a valid point, but on the other hand the `WithNaturalOrder` makes it pretty clear what happens. 

I don't think that `create(Iterable)` really conveys its semantics well when used: Is it really obvious that `SkipList.create(foos)` returns a collection that contains the same as `foos`?

So maybe we find a third alternative?

lembergerth replied:
How about `create(Iterable)` to create with natural order,
and `createOfSorted(SortedSet)` or `createWithSameOrder(SortedSet)`?

Edit:
I think the behavior of `create` is not clear by the name itself, but through the Java context.
Collections of the JDK (e.g., `ArrayList(Collection)` constructor ) as well as of Guava (e.g., `HashMultiset.create(Iterable)` factory method)
create a collection object containing the elements of the given Collection/Iterable, not just a view of it.

PhilippWendler replied:
I would agree. Is it clear that `createWithSameOrder()` also copies the content and not only the order?

lembergerth replied:
Good point.

I would prefer `copyOfWithSameOrder`, similar to what I used before, but that would be even less consistent with the Guava convention.

Even though it is less than optimal, I think we can also assume that the parameter implies that the elements of the given list will be copied (otherwise, why not use a `Comparator`-parameter)?
But maybe you have some other good idea for the naming.

PhilippWendler commented on src/org/sosy_lab/common/collect/SkipList.java:
```diff
@@ -163,37 +166,86 @@ private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFound
+    return new SkipList<>(pCollection);
+  }
+
+  public static <T> SkipList<T> copyOf(SortedSet<T> pSortedSet) {
```
How about `copyOfWithSameOrder`?

Or do you think the behavior of this method is what most people would expect from a `copyOf` method with no further explanation?

I just noted that Guava's `ImmutableSortedSet` has the reverse naming: `copyOf` uses natural order, `copyOfSorted` uses the same order. This is short and easy, and I think would be nice to adopt.

lembergerth replied:
Here I used `copyOf` because the elements and their order are both kept, so it is a real 'copy' (so also 'yes' to your second question).

But for factory methods, Guava actually seems to differentiate between `ImmutableXXX` and other collections (e.g., `TreeMultiset`). The second only provide `create(...)` methods.

So should we just align ourselves with the `ImmutableXXX` classes for all factory methods, to have a naming scheme consistent with Guava classes, since you mentioned that `of(…)` methods are useful?

lembergerth committed:
Remove trailing whitespace

lembergerth committed:
Format with google formater

lembergerth committed:
SkipList is final

lembergerth committed:
Rename SkipList#create... methods

lembergerth committed:
Remove unnecessary return param check in SkipListTest

lembergerth committed:
Format SkipList

lembergerth commented:
Experiments have shown that a naive implementation of the `OrderStatisticsSet` with a `TreeMap` and `Iterables#get` shows performance comparable to the SkipList implementation, for waitlists the size we encounter them in sv-benchmark tasks. So for now, this implementation is unnecessary.
